File: AddOneRowToTree/fastestSolution.java
// link: https://leetcode.com/problems/add-one-row-to-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int v, int d) {
        if(d == 1){
            TreeNode a = new TreeNode(v);
            a.left = root;
            return a;
        }
        help(root , v , d);
        return root;
    }
    
    void help(TreeNode root , int v , int d){
        if(root == null){
            return ;
        }
        if(d-1 == 1){
            TreeNode a1 = new TreeNode(v);
            TreeNode a2 = new TreeNode(v);
            TreeNode  s = root.left;
            TreeNode s1 = root.right;
            root.left = a1;
            a1.left = s;
            root.right = a2;
            a2.right = s1;
            return ;
        }
        help(root.left  , v , d-1);
        help(root.right , v ,d-1);
    }
}

File: convertSortedListToBST/fastestSolution.cpp
// link: https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
	auto sortedListToBST(ListNode* head) -> TreeNode * {
		vector<int> vals;
		
		auto node = head;
		while (node) {
			vals.push_back(node->val);
			node = node->next;
		}

		return helper(vals, 0, vals.size() - 1);
	}

	auto helper(const vector<int>& vals, int l, int r) -> TreeNode * {
		if (l > r)
			return nullptr;

		int m = (l + r)/2;
		auto node = new TreeNode(vals[m]);

		if (l < r) {
			node->left = helper(vals, l, m-1);
			node->right = helper(vals, m+1, r);
		}

		return node;
	}
};